### PR TITLE
Stabilize `#![feature(target_feature_11)]`

### PR DESCRIPTION
**Pull request by LeSeulArtichaut**
### Summary

Allows for safe functions to be marked with `#[target_feature]` attributes.

Functions marked with `#[target_feature]` are generally considered as unsafe functions: they are unsafe to call, cannot be assigned to safe function pointers, and don't implement the `Fn*` traits.

However, calling them from other `#[target_feature]` functions with a superset of features is safe.

```rust
// Demonstration function
#[target_feature(enable = "avx2")]
fn avx2() {}

fn foo() {
    // Calling `avx2` here is unsafe, as we must ensure
    // that AVX is available first.
    unsafe {
        avx2();
    }
}

#[target_feature(enable = "avx2")]
fn bar() {
    // Calling `avx2` here is safe.
    avx2();
}
```

### Test cases

Tests for this feature can be found in [`src/test/ui/rfcs/rfc-2396-target_feature-11/`](https://github.com/rust-lang/rust/tree/b67ba9ba208ac918228a18321fc3a11a99b1c62b/src/test/ui/rfcs/rfc-2396-target_feature-11/).

### Edge cases

- https://github.com/rust-lang/rust/issues/73631

Closures defined inside functions marked with `#[target_feature]` inherit the target features of their parent function. They can still be assigned to safe function pointers and implement the appropriate `Fn*` traits.

```rust
#[target_feature(enable = "avx2")]
fn qux() {
    let my_closure = || avx2(); // this call to `avx2` is safe
    let f: fn() = my_closure;
}
```

This means that in order to call a function with `#[target_feature]`, you must show that the target-feature is available while the function executes *and* for as long as whatever may escape from that function lives.

### Documentation

- Reference: https://github.com/rust-lang/reference/pull/1181

---
cc tracking issue #69098
r? @ghost

### CODE REVIEW

**Comment by LeSeulArtichaut** [comment]
@rusbot label T-lang needs-fcp

**Comment by joshtriplett** [comment]
@rfcbot merge

**Comment by LeSeulArtichaut** [comment]
Oops, I should probably have removed the `T-compiler` label? Sorry for the inconvenience

**Comment by oli-obk** [comment]
@rfcbot cancel

**Comment by LeSeulArtichaut** [comment]
Tried to write up a stabilization report, feedback would be appreciated.

**Comment by lcnr** [comment]
how does this interact with type alias impl trait
```rust
#![feature(target_feature_11, type_alias_impl_trait)]

type Foo = impl FnOnce() + Default;

#[target_feature(enable = "avx2")]
fn avx2() {}

#[target_feature(enable = "avx2")]
fn qux() {
    let x: Foo = || avx2(); // currently errors cause the closure isn't `Default`.
}
```
in #77688 it was decided against adding `Default` for function types and closures but that decision wasn't "we will never add that impl in the future" so I do want something apart from closures aren't `Default` from preventing the above example to work

**Comment by LeSeulArtichaut** [comment]
@lcnr I think this is analogous to something like this?

```rust
unsafe fn unsf() {}

// This has some safety requirements...
unsafe fn foo() {
    // ...which might be used in the closure here
    let x: Foo = || unsf();
}
```

I think calling `x` through `Foo::default()` would be unsound?

Or something like

```rust
fn bar() {
    // check that some invariants are upheld, returning if not
    if !invariant_upheld { return; }

    let x: Foo = || unsafe { unsf(); };
}
```

Wouldn't it be surprising to be able to call `x` through `Default`?
    

This was raised in https://github.com/rust-lang/rust/pull/77688#issuecomment-725538587 but I couldn't find a proposed solution/workaround. If there is one, would it be applicable here as well?

**Comment by lcnr** [comment]
While `|| unsafe { some_unsafe_fn() }` is now a safe function pointer which we could leak this way, we still need `unsafe` somewhere to skip that boundary.

With target features the skip can be fully implicit (except for a `#[target_feature(...)]` attribute on the function constraining the opaque type).

> Wouldn't it be surprising to be able to call `x` through `Default`?

as `Foo` would need an explicit `Default` bound, I think no ^^

I realized another issue with implementing `Default` for closures automatically. In the future, we may uncover that an opaque type implements `Default` using specialization, changing implicit impls of `Default` to change currently sound APIs to be unsound, making this a breaking change.

I guess I am fine with stabilizing this then ^^ while it seems like a final nail in the coffin of `Default` for closures and function definitions, there are already a few other nails.

**Comment by est31** [comment]
:wave: Hello, I'm writing this comment in this **stabilization PR** to notify you, the authors of this PR, that https://github.com/rust-lang/rust/pull/100591 has been merged, which implemented a change in how features are stabilized.

Your PR has been filed before the change, so will likely require modifications in order to comply with the new rules. I recommend you to:

1. rebase the PR onto latest master, so that uses of the placeholder are possible.
2. replace the version numbers in the PR with the placeholder `CURRENT_RUSTC_VERSION`. For language changes, this means the version numbers in `accepted.rs` (example: 4caedbae869855a19ca49455e2cd338ae3b8c1ef). For library changes, this means the `since` fields (example e576a9b554f24a26c36c04b41c901636d190605b).

That's it! The `CURRENT_RUSTC_VERSION` placeholder will, as part of the release process, be replaced with the version number that the PR merged for. It can be used anywhere in `rust-lang/rust`, not just accepted.rs and the since fields.

If you have any questions, feel free to drop by the [zulip stream](https://rust-lang.zulipchat.com/#narrow/stream/241545-t-release/topic/libs.20stabilization.20placeholder), or ping me directly in this PR's thread. Thanks! :wave:

**Comment by LeSeulArtichaut** [comment]
Done! Thanks for the heads-up @est31

**Comment by Dylan-DPC** [comment]
@LeSeulArtichaut what's the status of this?

**Comment by joshtriplett** [comment]
@rfcbot merge

**Comment by nikomatsakis** [comment]
@rfcbot resolved

**Comment by LeSeulArtichaut** [comment]
> @rfcbot resolved

@nikomatsakis I'm guessing you meant `@rfcbot reviewed`?

**Comment by LeSeulArtichaut** [comment]
r? compiler

**Comment by nagisa** [comment]
r? rust-lang/compiler

File: compiler/rustc_feature/src/accepted.rs [hunk 1]
```diff
@@ -312,6 +312,8 @@ declare_features! (
     (accepted, struct_variant, "1.0.0", None, None),
     /// Allows `#[target_feature(...)]`.
     (accepted, target_feature, "1.27.0", None, None),
+    /// Allows the use of `#[target_feature]` on safe functions.
+    (accepted, target_feature_11, "CURRENT_RUSTC_VERSION", Some(69098), None),
```
Comment (estebank) [inline compiler/rustc_feature/src/accepted.rs:316]: Let's not forget to update this.

**Review by estebank — COMMENTED** [review]
r=me

File: compiler/rustc_feature/src/accepted.rs (hunk 1 above)
Comment (LeSeulArtichaut) [reply compiler/rustc_feature/src/accepted.rs:316]: This is now done when beta branches (see #100591)

**Comment by Nilstrieb** [comment]
The comment was resolved and @estebank approved it, so
@bors r=estebank

**Comment by workingjubilee** [comment]
...Hm. I was not aware this had been FCP'd. We are aware that this makes downgrading feature levels a backwards incompatible change for targets, right?

**Comment by est31** [comment]
@workingjubilee 

>  this makes downgrading feature levels a backwards incompatible change for targets

You mean when proposals like [this](https://github.com/rust-lang/compiler-team/issues/548) get accepted, the "default" set of default-enabled target features shrinks and some code calling `#[target_feature]` enabled attributes that were before part of the default set starts to fail compiling?

From testing the implementation, that seems to have been avoided by not having a default set of target features. Even if the target feature is enabled by default, code still must use unsafe. I wasn't involved much during implementation of this RFC so I can't comment on how intentional that was, but I vaguely remember a comment that they intentionally made the most restricted version, where there is also no implication between features that came after another, e.g. sse2 -> sse.

Take this code ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=5e662cf03e5b89eface12888233d97c6)):

```Rust
#![feature(target_feature_11)]

#[target_feature(enable = "sse")] fn foo() {}

#[target_feature(enable = "sse")] fn bar() { foo(); }

fn main() {
    #[cfg(target_feature = "sse")]
    bar(); //~ERROR call to function with `#[target_feature]` is unsafe and requires unsafe function or block
}
```

It will print the error as indicated. The test is written in a way that shows that `sse` is enabled by default (as is required on AMD 64 bit), and if it weren't, then it wouldn't error as the cfg would not expand. But it still errors even though `sse` is enabled.

Also note that downgrading of target requirements are breaking changes already, for example you could have code like:

```Rust
extern { fn fancy_fn_that_uses_sse(); }

fn main() {
    #[cfg(target_feature = "sse")]
    unsafe { fancy_fn_that_uses_sse(); }

    #[cfg(not(target_feature = "sse"))]
    compile_error!("sorry but we only have an sse using implementation of this!");
}
```

It doesn't use `target_feature_11` at all, but it would fail to compile upon a removal of `sse` from the default set of enabled target features.

In these scenarios I also think that compiler errors are preferable to runtime SIGILLs or worse.

One could think about extending the `target_feature` feature to also take the default-enabled target features into account, but that should be discussed in other threads, not here.
